File: python/scripts/build_single_schema.py
import json

from ngff_rfc8.load_schemas import _get_list_schema_files


def _replace_refs(obj):
    for key, value in obj.items():
        if key == "$ref" and not value.startswith("#"):
            id = value[: -len(".schema")]
            obj[key] = f"#/$defs/{id}"
        elif type(value) is dict:
            _replace_refs(value)
        elif type(value) is list:
            for v in value:
                if type(v) is dict:
                    _replace_refs(v)


def main():
    schemas = {}
    root_schema = None
    defs = {}
    for path in _get_list_schema_files():
        schema = json.loads(path.read_text())
        if schema["$id"] == "ome.schema":
            root_schema = schema
        else:
            if "$defs" in schema:
                d = schema["$defs"]
                _replace_refs(d)
                defs.update(d)
                del schema["$defs"]
            schemas[schema["$id"]] = schema
        del schema["$id"]

    _replace_refs(root_schema)

    for id, schema in schemas.items():
        id = id[: -len(".schema")]
        _replace_refs(schema)
        defs[id] = schema

    root_schema["$defs"] = defs

    print(json.dumps(root_schema, indent=2))


if __name__ == "__main__":
    main()

File: javascript/build-single-schema.js
import fs from 'fs';
import path from 'path';

const __dirname = import.meta.dirname;

const schemaFolder = path.resolve(__dirname, '..', 'schemas');
const targetFile = path.resolve(__dirname, 'ngff-rfc8-schema.json');

function replaceRefs(obj) {
  for (const [key, value] of Object.entries(obj)) {
    if (key === '$ref' && typeof value === 'string' && !value.startsWith('#')) {
      const id = value.replace('.schema', '');
      obj[key] = `#/$defs/${id}`;
    } else if (value !== null && typeof value === 'object') {
      if (Array.isArray(value)) {
        for (const v of value) {
          if (v !== null && typeof v === 'object') {
            replaceRefs(v);
          }
        }
      } else {
        replaceRefs(value);
      }
    }
  }
}

export function buildSingleSchema() {
  const schemas = {};
  let rootSchema = null;
  const defs = {};

  const files = fs
    .readdirSync(schemaFolder, { withFileTypes: true })
    .filter((item) => !item.isDirectory())
    .map((item) => path.resolve(schemaFolder, item.name));

  for (const filePath of files) {
    const schema = JSON.parse(fs.readFileSync(filePath, 'utf-8'));

    if (schema['$id'] === 'ome.schema') {
      rootSchema = schema;
    } else {
      if ('$defs' in schema) {
        const d = schema['$defs'];
        replaceRefs(d);
        Object.assign(defs, d);
        delete schema['$defs'];
      }
      schemas[schema['$id']] = schema;
    }
    delete schema['$id'];
  }

  replaceRefs(rootSchema);

  for (const [id, schema] of Object.entries(schemas)) {
    const cleanId = id.replace('.schema', '');
    replaceRefs(schema);
    defs[cleanId] = schema;
  }

  rootSchema['$defs'] = defs;

  fs.writeFileSync(targetFile, JSON.stringify(rootSchema, null, 2));
}
